File: ch07/CSharpServerControl1/Listing07-40.cs
using System.ComponentModel;
using System.Web.UI;
using System.Web.UI.WebControls;

namespace CSharpServerControl1
{
    [Designer(typeof(MultiRegionControlDesigner))]
    [ToolboxData("<{0}:Listing0740 runat=\"server\" width=\"100%\">" +
        "</{0}:Listing0740>")]
    public class Listing0740 : CompositeControl
    {

        // Define the templates that represent 2 views on the control
        private ITemplate _view1;
        private ITemplate _view2;

        // These properties are inner properties
        [PersistenceMode(PersistenceMode.InnerProperty), DefaultValue(null)]
        public virtual ITemplate View1
        {
            get { return _view1; }
            set { _view1 = value; }
        }

        [PersistenceMode(PersistenceMode.InnerProperty), DefaultValue(null)]
        public virtual ITemplate View2
        {
            get { return _view2; }
            set { _view2 = value; }
        }

        // The current view on the control; 0= view1, 1=view2, 2=all views
        private int _currentView = 0;
        public int CurrentView
        {
            get { return _currentView; }
            set { _currentView = value; }
        }

        protected override void CreateChildControls()
        {
            Controls.Clear();

            ITemplate template = View1;
            if (_currentView == 1)
                template = View2;

            Panel p = new Panel();
            Controls.Add(p);

            if (template != null)
                template.InstantiateIn(p);
        }
    }
}

File: ch07/VbServerControl1/Listing07-40.vb
Imports System.ComponentModel
Imports System.Web.UI
Imports System.Web.UI.WebControls

<Designer(GetType(MultiRegionControlDesigner))> _
<ToolboxData("<{0}:Listing0740 runat=server width=100%></{0}:Listing0740>")> _
Public Class Listing0740
    Inherits CompositeControl

    ' Define the templates that represent 2 views on the control
    Private _view1 As ITemplate
    Private _view2 As ITemplate

    ' These properties are inner properties
    <PersistenceMode(PersistenceMode.InnerProperty), DefaultValue("")> _
    Public Overridable Property View1() As ITemplate
        Get
            Return _view1
        End Get
        Set(ByVal value As ITemplate)
            _view1 = value
        End Set
    End Property

    <PersistenceMode(PersistenceMode.InnerProperty), DefaultValue("")> _
    Public Overridable Property View2() As ITemplate
        Get
            Return _view2
        End Get
        Set(ByVal value As ITemplate)
            _view2 = value
        End Set
    End Property

    ' The current view on the control; 0= view1, 1=view2, 2=all views
    Private _currentView As Int32 = 0
    Public Property CurrentView() As Int32
        Get
            Return _currentView
        End Get
        Set(ByVal value As Int32)
            _currentView = value
        End Set
    End Property

    Protected Overrides Sub CreateChildControls()
        MyBase.CreateChildControls()

        Controls.Clear()

        Dim template As ITemplate = View1
        If (_currentView = 1) Then
            template = View2
        End If

        Dim p As New Panel()
        Controls.Add(p)

        If (Not template Is Nothing) Then
            template.InstantiateIn(p)
        End If
    End Sub

End Class